### PR TITLE
Adding function: Allow users to designate the number of batches

### PR DESCRIPTION
**Pull request by Unous1996**
Now when calling the register_model function, the users can designate the batch size so that all the containers could use this specific batch size instead of using the default batching strategy. For example:
`connection.register_model(name="model_name", version="model_version15", input_type="floats",batch_size=5)`
The default value of the parameter "batch_size" is -1. Clipper will apply the default strategy under this situation.

### CODE REVIEW

**Comment by Unous1996** [comment]
OK to test.

**Comment by dcrankshaw** [comment]
jenkins ok to test

**Comment by dcrankshaw** [comment]
@Unous1996 There's an issue with the jenkins cluster right now, it's not your code that's causing the tests to fail. Hang tight, working to get the issue resolved.

**Comment by Unous1996** [comment]
OK. So are there anything I need to do right now?

File: clipper_admin/clipper_admin/clipper_admin.py [hunk 1]
```diff
@@ -278,7 +279,10 @@ def build_and_deploy_model(self,
             The number of replicas of the model to create. The number of replicas
             for a model can be changed at any time with
             :py:meth:`clipper.ClipperConnection.set_num_replicas`.
-
+        batch_size : int, optional
+            The user defined batch size.
+            The default value is -1, which means that the user did not define a specific batch size.
```
Comment (Corey-Zumar) [inline clipper_admin/clipper_admin/clipper_admin.py]: Replace lines 284-285 with: "If the default value of -1 is used, Clipper will adaptively calculate the batch size for individual replicas of this model."

File: clipper_admin/clipper_admin/clipper_admin.py [hunk 2]
```diff
@@ -278,7 +279,10 @@ def build_and_deploy_model(self,
             The number of replicas of the model to create. The number of replicas
             for a model can be changed at any time with
             :py:meth:`clipper.ClipperConnection.set_num_replicas`.
-
+        batch_size : int, optional
+            The user defined batch size.
```
Comment (Corey-Zumar) [inline clipper_admin/clipper_admin/clipper_admin.py]: user defined >> user-defined

File: clipper_admin/clipper_admin/clipper_admin.py [hunk 3]
```diff
@@ -331,7 +336,10 @@ def build_model(self,
             The Docker container registry to push the freshly built model to. Note
             that if you are running Clipper on Kubernetes, this registry must be accesible
             to the Kubernetes cluster in order to fetch the container from the registry.
-
+        batch_size : int, optional
+            The user defined batch size.
+            The default value is -1, which means that the user did not define a specific batch size.
```
Comment (Corey-Zumar) [inline clipper_admin/clipper_admin/clipper_admin.py]: Update the documentation for this method in the same way as for `build_and_deploy_model`.

File: clipper_admin/clipper_admin/clipper_admin.py [hunk 4]
```diff
@@ -436,6 +445,10 @@ def deploy_model(self,
             The number of replicas of the model to create. The number of replicas
             for a model can be changed at any time with
             :py:meth:`clipper.ClipperConnection.set_num_replicas`.
+        batch_size : int, optional
+            The user defined batch size.
+            The default value is -1, which means that the user did not define a specific batch size.
```
Comment (Corey-Zumar) [inline clipper_admin/clipper_admin/clipper_admin.py]: Update the documentation for this method in the same way as for `build_and_deploy_model`.

File: clipper_admin/clipper_admin/clipper_admin.py [hunk 5]
```diff
@@ -453,10 +466,15 @@ def deploy_model(self,
...
         version = str(version)
         _validate_versioned_model_name(name, version)
+        '''
         self.cm.deploy_model(
-            name, version, input_type, image, num_replicas=num_replicas)
+            name, version, input_type, batch_size=batch_size, image, num_replicas=num_replicas)
+        '''
+        ''' 
+        '''
+        self.cm.deploy_model(name=name, version=version, input_type=input_type, image=image, batch_size=batch_size)
```
Comment (Corey-Zumar) [inline clipper_admin/clipper_admin/clipper_admin.py]: add the `num_replicas` keyword argument

File: clipper_admin/clipper_admin/clipper_admin.py [hunk 6]
```diff
@@ -453,10 +466,15 @@ def deploy_model(self,
             raise UnconnectedException()
         version = str(version)
         _validate_versioned_model_name(name, version)
+        '''
         self.cm.deploy_model(
-            name, version, input_type, image, num_replicas=num_replicas)
+            name, version, input_type, batch_size=batch_size, image, num_replicas=num_replicas)
```
Comment (Corey-Zumar) [inline clipper_admin/clipper_admin/clipper_admin.py]: Remove this duplicate call to `deploy_model`.

File: clipper_admin/clipper_admin/clipper_admin.py [hunk 7]
```diff
@@ -453,10 +466,15 @@ def deploy_model(self,
...
+        '''
+        '''
```
Comment (Corey-Zumar) [inline clipper_admin/clipper_admin/clipper_admin.py]: Remove these extraneous " ''' " lines

File: clipper_admin/clipper_admin/clipper_admin.py [hunk 8]
```diff
@@ -495,6 +514,10 @@ def register_model(self,
         labels : list(str), optional
             A list of strings annotating the model. These are ignored by Clipper
             and used purely for user annotations.
+        batch_size : int, optional
+            The user defined batch size.
+            The default value is -1, which means that the user did not define a specific batch size.
```
Comment (Corey-Zumar) [inline clipper_admin/clipper_admin/clipper_admin.py]: Update the documentation for this method in the same way as for `build_and_deploy_model`.

File: containers/python/noop_container.py [hunk 9]
```diff
@@ -3,13 +3,15 @@
 import os
```
Comment (Corey-Zumar) [inline containers/python/noop_container.py]: Revert the changes to this file. These changes were only made for local testing purposes.

File: examples/basic_query/example_client.py [hunk 10]
```diff
@@ -36,8 +36,8 @@ def signal_handler(signal, frame):
 if __name__ == '__main__':
```
Comment (Corey-Zumar) [inline examples/basic_query/example_client.py]: Revert the changes to this file

File: src/libclipper/include/clipper/redis.hpp [hunk 11]
```diff
@@ -118,6 +118,8 @@ boost::optional<std::string> get_current_model_version(
  *
  * \param container_name should be the name of a Docker container
  * \param model_data_path should be the path on the Clipper host
+ * \param designated_batch_size should be the user defined batch size of the container, it's default value(-1)
```
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/redis.hpp]: change `designated_batch_size` to `batch_size`.

Comment (Corey-Zumar) [inline src/libclipper/include/clipper/redis.hpp]: Change the description to "should be the user-defined batch size for all replicas of the model. Its default value (-1) indicates that the batch size was unspecified and that Clipper should calculate it adaptively.

File: src/libclipper/include/clipper/redis.hpp [hunk 12]
```diff
@@ -169,6 +171,13 @@ std::vector<std::string> get_model_versions(redox::Redox& redis,
  * \return Returns a vector of model names as strings. If no model names
  * were found, then an empty vector will be returned.
  */
+std::unordered_map<std::string,std::string> get_model_by_key(redox::Redox& redis, const std::string & key);
+ /**
+  * get_model_by_key
+  *
+  * @param redis
+  * @return
```
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/redis.hpp]: Add parameter and return type documentation to this method

File: src/libclipper/include/clipper/task_executor.hpp [hunk 13]
```diff
@@ -272,6 +287,16 @@ class TaskExecutor {
             VersionedModelId vm = VersionedModelId(
                 container_info["model_name"], container_info["model_version"]);
             int replica_id = std::stoi(container_info["model_replica_id"]);
+            //#int batch_size = std::stoi(container_info["batch_size"]);
+            boost::optional<int> batch_size = -1;
+            for( auto it = container_info.begin(); it != container_info.end(); it++){
```
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/task_executor.hpp]: Instead of iterating over the key-value pairs in the `container_info` map, use the [find](http://en.cppreference.com/w/cpp/container/unordered_map/find) function and check the validity result of the "find" operation as:

```
auto &batch_size_search = container_info.find("batch_size")
if(batch_size_search != container_info.end()) {
    // The entry exists, obtain the value and update the batch size
    ...
}
```

File: src/libclipper/src/containers.cpp [hunk 14]
```diff
@@ -76,6 +96,16 @@ double ModelContainer::get_average_throughput_per_millisecond() {
 }
 
 size_t ModelContainer::get_batch_size(Deadline deadline) {
+  std::cout<<"this->batch_size="<<this->batch_size_<<std::endl;
```
Comment (Corey-Zumar) [inline src/libclipper/src/containers.cpp]: Remove print statements

File: src/libclipper/src/containers.cpp [hunk 15]
```diff
@@ -18,13 +18,33 @@
 namespace clipper {
 
 const std::string LOGGING_TAG_CONTAINERS = "CONTAINERS";
-
+/*
```
Comment (Corey-Zumar) [inline src/libclipper/src/containers.cpp]: Remove commented code

File: src/libclipper/include/clipper/containers.hpp [hunk 16]
```diff
@@ -21,7 +21,9 @@ class ModelContainer {
  public:
   ~ModelContainer() = default;
   ModelContainer(VersionedModelId model, int container_id, int replica_id,
-                 InputType input_type);
+                  InputType input_type);
+  ModelContainer(VersionedModelId model, int container_id, int replica_id,
+                 InputType input_type, boost::optional<int> batch_size);
```
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/containers.hpp]: If we're always specifying `batch_size` as either `-1` or a positive integer, we don't need to use an optional here. I had originally suggested passing an absent optional value to the `ModelContainer` constructor if the user has not defined the batch size, but using `-1` directly seems fine. Change this parameter type and the `batch_size_` class member type to `int`.

File: src/libclipper/src/containers.cpp [hunk 17]
```diff
@@ -108,8 +138,17 @@ void ActiveContainers::add_container(VersionedModelId model, int connection_id,
                      model.get_name(), model.get_id(), connection_id,
                      replica_id, get_readable_input_type(input_type));
   boost::unique_lock<boost::shared_mutex> l{m_};
+
+  // Set a default batch size of 1
+  int batch_size = 1;
```
Comment (Corey-Zumar) [inline src/libclipper/src/containers.cpp]: The default value should be `-1` if the user has not defined a batch size for the container's associated model.

File: src/libclipper/src/containers.cpp [hunk 18]
```diff
@@ -108,8 +138,17 @@ void ActiveContainers::add_container(VersionedModelId model, int connection_id,
...
+  auto batch_size_search = batch_sizes_.find(model);
+  if(batch_size_search != batch_sizes_.end()) {
+    std::cout<<"MODEL_FOUND"<<std::endl;
```
Comment (Corey-Zumar) [inline src/libclipper/src/containers.cpp]: Remove print statement

File: src/management/src/management_frontend.hpp [hunk 19]
```diff
@@ -670,6 +670,7 @@ class RequestHandler {
     InputType input_type = clipper::parse_input_type(input_type_raw);
     std::string container_name = get_string(d, "container_name");
     std::string model_data_path = get_string(d, "model_data_path");
+    boost::optional<int> batch_size = get_int(d, "batch_size");
```
Comment (Corey-Zumar) [inline src/management/src/management_frontend.hpp]: Add a check to ensure that `batch_size` is either `-1` or a positive integer, and throw a `ManagementOperationError` if this condition is not met.

File: src/libclipper/include/clipper/task_executor.hpp [hunk 20]
```diff
@@ -260,6 +260,21 @@ class TaskExecutor {
...
+
+    redis::subscribe_to_model_changes(redis_subscriber_,
+                    [this, task_executor_valid = active_ ](const std::string &key,
+                                                           const std::string &event_type) {
+                log_info(LOGGING_TAG_TASK_EXECUTOR,"subscribe_to_model_changes is called");
+                if (event_type == "hset" && *task_executor_valid) {
+                    auto model_info = clipper::redis::get_model_by_key(redis_connection_, key);
+                    VersionedModelId model_id = VersionedModelId(model_info["model_name"], model_info["model_version"]);
+                    int batch_size = std::stoi(model_info["batch_size"]);
+                    std::cout<<"registered_batch_size"<<batch_size<<std::endl;
```
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/task_executor.hpp]: Replace this with `log_info_formatted("Registered batch size of {} for model {}:{}", batch_size, model_id.get_name(), model_id.get_id())`

File: src/libclipper/include/clipper/task_executor.hpp [hunk 21]
```diff
@@ -272,6 +287,16 @@ class TaskExecutor {
             VersionedModelId vm = VersionedModelId(
                 container_info["model_name"], container_info["model_version"]);
             int replica_id = std::stoi(container_info["model_replica_id"]);
+            //#int batch_size = std::stoi(container_info["batch_size"]);
+            boost::optional<int> batch_size = -1;
+            for( auto it = container_info.begin(); it != container_info.end(); it++){
+              std::cout<<"it->first="<<it->first<<std::endl;
```
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/task_executor.hpp]: Remove these print statements

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]

**Comment by Corey-Zumar** [comment]
@Unous1996 This is definitely on the right track! Most of the review comments are concerned with code cleanup and readability

File: src/management/src/management_frontend.hpp (hunk 19 above)
Comment (Unous1996) [reply src/management/src/management_frontend.hpp]: OK. I will change it as soon as possible.

**Comment by Unous1996** [comment]
Is there anything wrong with the cluster right now? Or there's some issue with my code.

File: src/management/src/management_frontend.hpp [hunk 22]
```diff
@@ -670,7 +670,14 @@ class RequestHandler {
     InputType input_type = clipper::parse_input_type(input_type_raw);
     std::string container_name = get_string(d, "container_name");
     std::string model_data_path = get_string(d, "model_data_path");
+    int batch_size = get_int(d, "batch_size");
 
+    // The barch_size should be either positive or -1
+    if( batch_size<0 && batch_size!= -1){
```
Comment (Corey-Zumar) [inline src/management/src/management_frontend.hpp]: This condition doesn't handle the invalid case where `batch_size` is zero

File: src/libclipper/src/containers.cpp [hunk 23]
```diff
@@ -76,6 +77,10 @@ double ModelContainer::get_average_throughput_per_millisecond() {
 }
 
 size_t ModelContainer::get_batch_size(Deadline deadline) {
+  if (this->batch_size_ != -1){
```
Comment (Corey-Zumar) [inline src/libclipper/src/containers.cpp]: There's no need to dereference `this` here. Just refer to `batch_size`

File: src/management/src/management_frontend.hpp [hunk 24]
```diff
@@ -670,7 +670,14 @@ class RequestHandler {
     InputType input_type = clipper::parse_input_type(input_type_raw);
     std::string container_name = get_string(d, "container_name");
     std::string model_data_path = get_string(d, "model_data_path");
+    int batch_size = get_int(d, "batch_size");
 
+    // The barch_size should be either positive or -1
```
Comment (Corey-Zumar) [inline src/management/src/management_frontend.hpp]: Fix typo: "barch_size" >> "batch_size"

File: containers/python/noop_container.py (hunk 9 above)
Comment (Corey-Zumar) [reply containers/python/noop_container.py]: Revert this file

File: examples/basic_query/example_client.py [hunk 25]
```diff
@@ -45,4 +45,4 @@ def signal_handler(signal, frame):
             predict(clipper_conn.get_query_addr(), np.random.random(200))
             time.sleep(0.2)
     except Exception as e:
-        clipper_conn.stop_all()
+        clipper_conn.stop_all()
```
Comment (Corey-Zumar) [inline examples/basic_query/example_client.py:63]: Revert this file

File: src/libclipper/src/containers.cpp [hunk 26]
```diff
@@ -108,8 +113,16 @@ void ActiveContainers::add_container(VersionedModelId model, int connection_id,
                      model.get_name(), model.get_id(), connection_id,
                      replica_id, get_readable_input_type(input_type));
   boost::unique_lock<boost::shared_mutex> l{m_};
+
+  // Set a default batch size of -1
+  int batch_size = -1;
```
Comment (Corey-Zumar) [inline src/libclipper/src/containers.cpp]: Define a `constexpr DEFAULT_BATCH_SIZE = -1` constant in `container.hpp` and refer to it in all relevant locations, instead of using the value `-1` directly

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]
A few more comments

File: examples/basic_query/example_client.py (hunk 25 above)
Comment (Unous1996) [reply examples/basic_query/example_client.py:63]: It seems that I could not find the difference. Both seems to call clipper_conn.stop_all()

**Comment by dcrankshaw** [comment]
The tests are failing right because of the format checker. Run `./bin/format_code.sh` to format the code correctly.

**Comment by Unous1996** [comment]
OK. Thanks.

**Comment by Unous1996** [comment]
Well, it seemed that I had problem when running the format_code.sh

`/Users/tim/Documents/Clipper_proj/clipper/bin/..
xargs: clang-format: No such file or directory
xargs: clang-format: No such file or directory`

I have already install the clang format file, move it to the ./bin/directory and changed the folder name to "clang-format", but it did not work.

File: src/libclipper/include/clipper/task_executor.hpp [hunk 27]
```diff
@@ -272,9 +291,19 @@ class TaskExecutor {
...
             int replica_id = std::stoi(container_info["model_replica_id"]);
-            active_containers_->add_container(
-                vm, std::stoi(container_info["zmq_connection_id"]), replica_id,
-                parse_input_type(container_info["input_type"]));
+              active_containers_->add_container(
+                      vm, std::stoi(container_info["zmq_connection_id"]), replica_id,
+                      parse_input_type(container_info["input_type"]));
+
+            auto model_info = redis::get_model(redis_connection_, vm);
+            VersionedModelId model_id = VersionedModelId(model_info["model_name"], model_info["model_version"]);
```
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/task_executor.hpp]: The value of this variable will be identical to `vm` (defined on line 291) and can be deleted

File: src/libclipper/include/clipper/task_executor.hpp [hunk 28]
```diff
@@ -272,9 +291,19 @@ class TaskExecutor {
...
+              int batch_size = DEFAULT_BATCH_SIZE;
+              auto batch_size_search = model_info.find("batch_size");
+              if(batch_size_search != model_info.end()){
+                batch_size = std::stoi(model_info["batch_size"]);
+                active_containers_->register_batch_size(model_id, batch_size);
```
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/task_executor.hpp]: move  this outside of the `if` statement and replace `model_id` with `vm`

File: src/libclipper/include/clipper/task_executor.hpp [hunk 29]
```diff
@@ -272,9 +291,19 @@ class TaskExecutor {
...
+              active_containers_->add_container(
+                      vm, std::stoi(container_info["zmq_connection_id"]), replica_id,
+                      parse_input_type(container_info["input_type"]));
+
+            auto model_info = redis::get_model(redis_connection_, vm);
+            VersionedModelId model_id = VersionedModelId(model_info["model_name"], model_info["model_version"]);
+              int batch_size = DEFAULT_BATCH_SIZE;
+              auto batch_size_search = model_info.find("batch_size");
+              if(batch_size_search != model_info.end()){
+                batch_size = std::stoi(model_info["batch_size"]);
```
Comment (Corey-Zumar) [inline src/libclipper/include/clipper/task_executor.hpp]: Replace with: `batch_size = std::stoi(batch_size_search->second)`

**Review by Corey-Zumar — CHANGES_REQUESTED** [review]
@Unous1996 This just needs a few more changes

**Comment by Unous1996** [comment]
OK. I will try to fix it.

**Review by Corey-Zumar — APPROVED** [review]
LGTM. Will merge when code formatting completes
